### ISSUE TITLE
bug: Firefox Imperative and Declarative search engines conflict

### ISSUE THREAD

**Issue by cvoges12** [nix-community/home-manager#4243]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Steps to Reproduce

1. Declare search engines using `programs.firefox.profiles.<name>.search.engines` options (declarative).
2. Install search engine extension (imperative).
3. Try to use search engine extension.

*Note: this still breaks even if you installed the search engine extensions through home-manager using ``*

### Actual behavior

Only the search engines defined using the `programs.firefox.profiles.<name>.search.engines` options work correctly.

### Expected behavior

Both the search engines via extensions and the home-manager option should work as normal.

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 6.1.31, NixOS, 23.05 (Stoat), 23.05.20230711.8163a64`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by kira-bruneau** [comment]
Hmm, there's not really a good way of working around this since both will use the same file for configuring search engines, and Firefox doesn't offer some kind of config layering for search engines.

Most of the time in home manager you loose out on imperative configuration if you opt for a declarative configuration. I'd recommend porting those extension search engines to the declarative configuration if you want both.

I'll try to see if there's some way we can merge the config from extensions defined in `programs.firefox.profiles.<name>.extensions`, but that would require [IFDs](https://nixos.wiki/wiki/Import_From_Derivation#When_to_use_IFD).

**Comment by cvoges12** [comment]
Okay, I'm actually finding some *seemingly* non-deterministic behavior. I somehow managed to get some search extensions to work after a few restarts of firefox and running home-manager a few more times. I'm trying to investigate further, but I'm really clueless as to why.

I think I should also note that I'm using `search.force` for my configuration.

**Comment by kira-bruneau** [comment]
Oh, so sometimes Firefox will override the file and that can occur between home-manager activations, but the next time home-manager is activated, it will be reverted to the declarative configuration.

**Comment by cvoges12** [comment]
I don't think it's much of "reverting" per se. When the seach extensions do finally start working, the declarative configurations of the search engine still work as well. It's some how reading from both.

_**Note**: I have no experience with imperatively installing search extensions. These are extensions installed from home-manager itself._

**Comment by kira-bruneau** [comment]
Oh, but that would still make sense. Firefox would imperatively merge the configuration generated by home-manager with the extension configuration.

So when Firefox overrides the configuration, it would keep the declarative options, but when home-manager gets activated it would revert back to only the declarative configuration.

**Comment by cvoges12** [comment]
Yes, but, if I remember correctly, what also happens is that home-manager stops being able to manage firefox after firefox makes imperative changes such as installing extensions on its own.

**Comment by kira-bruneau** [comment]
If you're using the `force` option for the search engines then home-manager is going to forcibly replace the file rather than complain about the file already existing.

**Comment by kira-bruneau** [comment]
I feel like this is all pretty confusing because of the way Firefox handles search configuration. The extensions are installed declaratively, but Firefox imports the search configuration from those extensions imperatively 🫠

**Comment by cvoges12** [comment]
Okay, new update. Some of the one's I've set are getting removed randomly, ie - wikipedia (en) set to alias "w" just got removed today.
